dffml/model/model.py

```diff
@@ -80,28 +88,87 @@ def __init__(self, config):
         # TODO Just in case its a string. We should make it so that on
         # instantiation of an @config we convert properties to their correct
         # types.
-        directory = getattr(self.config, "directory", None)
-        if isinstance(directory, str):
-            directory = pathlib.Path(directory)
-        if isinstance(directory, pathlib.Path):
-            # to treat "~" as the the home directory rather than a literal
-            directory = directory.expanduser().resolve()
-            self.config.directory = directory
+        location = getattr(self.config, "location", None)
+        if isinstance(location, str):
+            location = pathlib.Path(location)
+        if isinstance(location, pathlib.Path):
+            # to treat "~" as the the home location rather than a literal
+            location = location.expanduser().resolve()
+            self.config.location = location
+        self.location_is_file = self.config.location.is_file()
 
     def __call__(self) -> ModelContext:
-        self._make_config_directory()
+        self._make_config_location()
         return self.CONTEXT(self)
 
-    def _make_config_directory(self):
+    async def __aenter__(self):
+        if self.location_is_file:
+            location = self.config.location
+            temp_dir = self._get_directory()
+            if zipfile.is_zipfile(location):
+                await self._run_operation("extract", location, temp_dir)
+            self.config.location = temp_dir
+            self.config.file_location = location
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        if self.location_is_file:
+            config_path = self.config.location / "config.json"
+            config_path.write_text(json.dumps(self.config))
+            output_location = self.config.file_location
+            await self._run_operation(
+                "archive", self.temp_dir, output_location
+            )
+            shutil.rmtree(self.temp_dir)
+
+    async def _run_operation(self, action, input_path, output_path):
+        operations = {
+            "zip": {
+                "extract": extract_zip_archive,
+                "archive": make_zip_archive,
+            }
+        }
+        file_type = (
+            input_path.suffix if action == "extract" else output_path.suffix
+        )
+        operation = operations[file_type][action]
+
+        dataflow = self._create_dataflow(
+            operation,
+            {
+                "input_directory_path": input_path,
+                "output_file_path": output_path,
+            },
+        )
+
+        async for _, _ in MemoryOrchestrator.run(dataflow):
```

Comment:
Let's use the `high_level:run` function

Reply:
I actually used that before but it was not able to import it in `model.py` file probably because of some circular dependency..... I was getting the error cannot import name run from high_level

Reply:
A lot of errors like this : 
```
ImportError: Failed to import test module: test_slr
Traceback (most recent call last):
  File "/usr/lib/python3.7/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
  File "/home/saahil/rps/dffml/tests/model/test_slr.py", line 4, in <module>
    from dffml import SLRModel, run_consoletest, AsyncTestCase
  File "/home/saahil/rps/dffml/dffml/__init__.py", line 69, in <module>
    for import_name, module in modules(root, package_name, skip=skip):
  File "/home/saahil/rps/dffml/dffml/__init__.py", line 30, in modules
    yield import_name, importlib.import_module(import_name)
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/home/saahil/rps/dffml/dffml/high_level.py", line 11, in <module>
    from .model.model import Model
  File "/home/saahil/rps/dffml/dffml/model/__init__.py", line 18, in <module>
    from .model import Model, ModelContext, SimpleModel, ModelNotTrained
  File "/home/saahil/rps/dffml/dffml/model/model.py", line 26, in <module>
    from ..high_level import run
ImportError: cannot import name 'run' from 'dffml.high_level' (/home/saahil/rps/dffml/dffml/high_level.py)
```

Reply:
I think it might be due to some circular dependency loop being formed .... 🤔